iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,153 @@
+            }
+            // There can't be more than one element
+            break;
+        }
+        index++;
+        iterator++;
+    }
+
+    auto removeIndexFromMap = [](std::multimap<capro::IdString_t, uint64_t>& map, uint64_t index) {
+        for (auto it = map.begin(); it != map.end();)
+        {
+            if (it->second == index)
+            {
+                it = map.erase(it);
+                continue;
+            }
+            else if (it->second > index)
+            {
+                // update index due to removed element
+                it->second--;
+            }
+            it++;
+        }
+    };
+
+    removeIndexFromMap(m_serviceMap, index);
+    removeIndexFromMap(m_instanceMap, index);
+
+    return removedElement;
 }
 
-void ServiceRegistry::find(InstanceSet_t& instances,
+void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const capro::IdString_t& service,
-                           const capro::IdString_t& instance) const
+                           const capro::IdString_t& instance) const noexcept
 {
-    if (instance == Wildcard)
+    cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
+
+    // Find (K1, K2)
+    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
+    if (instance != Wildcard && service != Wildcard)
     {
-        for (auto& instance : m_serviceMap[service].instanceSet)
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleServices;
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleInstances;
+
+        auto rangeServiceMap = m_serviceMap.equal_range(service);
+        for (auto entry = rangeServiceMap.first; entry != rangeServiceMap.second; ++entry)
+        {
+            possibleServices.push_back(entry->second);
+        }
+
+        auto rangeInstanceMap = m_instanceMap.equal_range(instance);
+        for (auto entry = rangeInstanceMap.first; entry != rangeInstanceMap.second; ++entry)
+        {
+            possibleInstances.push_back(entry->second);
+        }
+
+        ::std::set_intersection(possibleServices.begin(),
+                                possibleServices.end(),
+                                possibleInstances.begin(),
+                                possibleInstances.end(),
+                                ::std::back_inserter(intersection));
+
+        for (auto& value : intersection)
         {
-            instances.push_back(instance);
+            searchResult.push_back(m_serviceDescriptionVector[value]);
         }
     }
-    else
+    // Find (*, K2)
+    // O(log n + #result)
+    else if (service == Wildcard && instance != Wildcard)
+    {
+        auto range = m_instanceMap.equal_range(instance);
+        for (auto entry = range.first; entry != range.second; ++entry)
+        {
+            searchResult.push_back(m_serviceDescriptionVector[entry->second]);
+        }
+    }
+    // Find (K1, *)
+    // O(log n + #result)
```

Comment:
It is just `O(log n)`

Reply:
Nope, #result may depend on n superlogarithmically and then it would be wrong. Leave as is.